env-var-docs/parser-package/src/env_var_docs/parser.py

```diff
@@ -0,0 +1,524 @@
+
+    Optional args:
+      default: If the field is not required and not defined, default will be returned. If required is True and default is not None, a ValueError is raised.
+      checks: Custom checks that can produce additional ParseErrors. CheckFns are only run if key in obj and isinstance(obj[key], json_type).
+      return_type: The type returned, can be a complex type. If set, extract_fn must not be None.
+      extract_fn: A function that returns a return_type instance from obj. extract_fn is only run if:
+        1. key in obj and isinstance(obj[key], json_type).
+        2. No checks return any ParseErrors.
+    """
+    if json_type not in [str, bool, dict, list]:
+        raise ValueError(
+            f"'{json_type}' is not a valid json_type, valid types are [str, bool, dict, list]"
+        )
+    if required and default is not None:
+        raise ValueError("'default' must be None if 'required' is True")
+    if return_type is not None and extract_fn is None:
+        raise ValueError(
+            "'extract_fn' must be provided if 'return_type' is provided")
+    if extract_fn is not None and return_type is None:
+        raise ValueError(
+            "'return_type' must be provided if 'extract_fn' is provided")
+
+    value = None
+    errors = []
+
+    if key not in obj:
+        if required:
+            errors.append(
+                ParseError(parent_path, f"'{key}' is a required field"))
+        else:
+            value = default
+    else:
+        if not isinstance(obj[key], json_type):
+            # Printing json_type directly produces strings like "<class'str'>".
+            # To make the error messages more readable, use these strings instead.
+            if json_type == str:
+                t = "string"
+            if json_type == bool:
+                t = "bool"
+            if json_type == dict:
+                t = "object"
+            if json_type == list:
+                t = "list"
+            errors.append(ParseError(_path(parent_path, key), f"must be a {t}"))
+        else:
+            for check in checks:
+                errors.extend(check(parent_path, obj))
+            if len(errors) == 0:
+                value = obj[key]
+                if extract_fn is not None:
+                    value = extract_fn(obj)
+
+    return value, errors
+
+
+def _path(*args: str) -> str:
+    """Creates a path from segments. Segments can not be the empty string."""
+    if len(args) == 0:
+        raise ValueError("must have at least one argument")
+    for a in args:
+        if a == "":
+            raise ValueError("arguments should not be the empty string")
+
+    return ".".join(args)
+
+
+def _ensure_no_extra_fields(
+        parent_path: str, obj: UnparsedJSON,
+        valid_fields: list[str]) -> ParseErrors:
+    """Returns ParseErrors for any keys in obj that are not in valid_fields."""
+    errors = []
+    for key in list(obj):
+        if key not in valid_fields:
```

Comment:
Is there any case when we also have to check that there are no duplicates or does json handle that internally?

Reply:
I did not think about this.  If there are duplicate keys in an object, the json package uses the last-specified key: https://docs.python.org/3/library/json.html#repeated-names-within-an-object.

To make it easier for us to catch bugs due to this behavior, I added a object_pairs_hook parameter to json.load that raises an exception if it is passed an object with duplicate keys.